### ISSUE TITLE
Can CrossEncoderTrainer support IterableDataset?

### ISSUE THREAD

**Issue by enze5088** [huggingface/sentence-transformers#3674]
Can CrossEncoderTrainer support IterableDataset? SentenceTransformerTrainer already supports it, but the corresponding validation logic in CrossEncoderTrainer has not been updated accordingly.

**Comment by nursery42** [comment]
In [cross_encoder_trainer](https://github.com/huggingface/sentence-transformers/blob/main/sentence_transformers/cross_encoder/trainer.py#L106)
```
class CrossEncoderTrainer(SentenceTransformerTrainer):
    def __init__(
        self,
        model: CrossEncoder | None = None,
        args: CrossEncoderTrainingArguments | None = None,
        train_dataset: Dataset | DatasetDict | dict[str, Dataset] | None = None,
        eval_dataset: Dataset | DatasetDict | dict[str, Dataset] | None = None,
        loss: nn.Module
        | dict[str, nn.Module]
        | Callable[[CrossEncoder], torch.nn.Module]
        | dict[str, Callable[[CrossEncoder], torch.nn.Module]]
        | None = None,
        evaluator: SentenceEvaluator | list[SentenceEvaluator] | None = None,
        data_collator: DataCollator | None = None,
        tokenizer: PreTrainedTokenizerBase | Callable | None = None,
        model_init: Callable[[], CrossEncoder] | None = None,
        compute_metrics: Callable[[EvalPrediction], dict] | None = None,
        callbacks: list[TrainerCallback] | None = None,
        optimizers: tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
        preprocess_logits_for_metrics: Callable[[torch.Tensor, torch.Tensor], torch.Tensor] | None = None,
    ) -> None:
```
but in [SentenceTransformerTrainer](https://github.com/huggingface/sentence-transformers/blob/main/sentence_transformers/trainer.py#L134)
```
train_dataset: Dataset | DatasetDict | IterableDataset | dict[str, Dataset] | None = None,
eval_dataset: Dataset | DatasetDict | IterableDataset | dict[str, Dataset] | None = None,
```

**Comment by tomaarsen** [comment]
Hello!

I'm afraid this is simply not possible right now. Sentence Transformers uses `transformers` behind the scenes, which uses `accelerate` again. When using `IterableDataset`, `accelerate` will concatenate batches from the IterableDataset, expecting every key-value pair after the data collator to only contain torch.Tensor values. However, the `CrossEncoderDataCollator` returns a dictionary with string values (expecting the tokenization to be done in the loss function). This will raise an error in `accelerate`.

So it's sadly not possible. I noticed that the docstring for the CE Trainer did still mention IterableDataset, so I've removed it there.

- Tom Aarsen
